chore: expose chain spec via commands

manav2401 opened the pull request:
Addresses https://github.com/paradigmxyz/reth/issues/14160. 

Exposed `chain_spec` in `Commands` where only the `NodeCommand` sub-command has the Chain arg while rest doesn't need any. For rest of the sub-commands, chain id (being used for setting up log dir) will be set as empty string. If you think I should use any other value as default, lmk.

mattsse commented on bin/reth/src/cli/mod.rs:
```diff
@@ -245,6 +233,22 @@ pub enum Commands<C: ChainSpecParser, Ext: clap::Args + fmt::Debug> {
     Prune(prune::PruneCommand<C>),
 }
 
+/// A provider for chain specifications.
+pub trait ChainSpecProvider<C: ChainSpecParser> {
+    /// Returns an optional chain specification.
+    fn chain_spec(&self) -> Option<&C::ChainSpec>;
+}
+
+impl<C: ChainSpecParser, Ext: clap::Args + fmt::Debug> ChainSpecProvider<C> for Commands<C, Ext> {
+    fn chain_spec(&self) -> Option<&C::ChainSpec> {
+        match self {
+            Commands::Node(cmd) => Some(&cmd.chain),
+            // For other subcommands that do not provide a chain, return None.
+            _ => None,
+        }
+    }
+}
```
I'd rather convert this to a native function of `Commands`

manav2401 replied:
Makes sense, done.

mattsse reviewed (changes requested):
great, smol suggestion

manav2401 commented:
@mattsse, I've addressed your comments. Just fyi, the docs failed earlier in CI as the chain arg was removed. I've updated the docs for the root command but haven't touched any sub-commands. Do you address them in the same pr or a separate one? Lmk, happy to do that. Thanks!

mattsse commented on bin/reth/src/cli/mod.rs:
```diff
@@ -245,6 +232,16 @@ pub enum Commands<C: ChainSpecParser, Ext: clap::Args + fmt::Debug> {
+    fn chain_spec(&self) -> Option<&C::ChainSpec> {
+        match self {
+            Self::Node(cmd) => Some(&cmd.chain),
+            // For other subcommands that do not provide a chain, return None.
+            _ => None,
```
we do have a few more instances where this is the case,

this will require a bit more copy paste work:

```
impl<C: ChainSpecParser> Command<C> {

    /// Returns the chainspec
    pub fn chain_spec(&self) -> &Arc<C::ChainSpec> {
        &self.env.chain
    }
}
```

basically on all commands the use a `<C>` generic

manav2401 replied:
Didn't know chain spec was buried inside `env`. Working on it, thanks!

mattsse commented on bin/reth/src/cli/mod.rs:
```diff
@@ -141,9 +128,9 @@ impl<C: ChainSpecParser<ChainSpec = ChainSpec>, Ext: clap::Args + fmt::Debug> Cl
-        self.logs.log_file_directory =
-            self.logs.log_file_directory.join(self.chain.chain.to_string());
+        // Add network name to logs dir (only set when using `Node` command)
+        let chain = self.command.chain_spec().map(|c| c.chain.to_string()).unwrap_or_default();
+        self.logs.log_file_directory = self.logs.log_file_directory.join(chain);
```
I'd like to use if let some here

mattsse reviewed (changes requested):
nice a few more things to do here

manav2401 commented:
@mattsse, have exposed `chain_spec` from all sub commands. Using `env` whenever possible to get the spec.
